merkle/log_verifier.go

```diff
@@ -0,0 +1,202 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package merkle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+// RootMismatchError occurs when an inclusion proof fails.
+type RootMismatchError struct {
+	ExpectedRoot   []byte
+	CalculatedRoot []byte
+}
+
+func (e RootMismatchError) Error() string {
+	return fmt.Sprintf("calculated root:\n%v\n does not match expected root:\n%v", e.CalculatedRoot, e.ExpectedRoot)
+}
+
+// LogVerifier verifies inclusion and consistency proofs for append only logs.
+type LogVerifier struct {
+	hasher TreeHasher
+}
+
+// NewLogVerifier returns a new LogVerifier for a tree.
+func NewLogVerifier(hasher TreeHasher) LogVerifier {
+	return LogVerifier{
+		hasher: hasher,
+	}
+}
+
+// VerifyInclusionProof verifies the correctness of the proof given the passed in information about the tree and leaf.
+func (m LogVerifier) VerifyInclusionProof(leafIndex, treeSize int64, proof [][]byte, root []byte, leaf []byte) error {
+	calcRoot, err := m.RootFromInclusionProof(leafIndex, treeSize, proof, leaf)
+	if err != nil {
+		return err
+	}
+	if !bytes.Equal(calcRoot, root) {
+		return RootMismatchError{
+			CalculatedRoot: calcRoot,
+			ExpectedRoot:   root,
+		}
+	}
+	return nil
+}
+
+// RootFromInclusionProof calculates the expected tree root given the proof and leaf.
+// leafIndex starts at 0. treeSize starts at 1.
+func (m LogVerifier) RootFromInclusionProof(leafIndex, treeSize int64, proof [][]byte, leaf []byte) ([]byte, error) {
+	if leafIndex >= treeSize {
+		return nil, fmt.Errorf("leafIndex %d > treeSize %d", leafIndex, treeSize)
+	}
+	if leafIndex < 0 || treeSize < 1 {
+		return nil, errors.New("leafIndex < 0 or treeSize < 1")
+	}
+
+	nodeIndex := leafIndex
+	lastNode := treeSize - 1
+	nodeHash := m.hasher.HashLeaf(leaf)
+	proofIndex := 0
+
+	for lastNode > 0 {
+		if proofIndex == len(proof) {
+			return nil, fmt.Errorf("insuficient number of proof components (%d) for treeSize %d", len(proof), treeSize)
+		}
+		if isRightChild(nodeIndex) {
+			nodeHash = m.hasher.HashChildren(proof[proofIndex], nodeHash)
+			proofIndex++
+		} else if nodeIndex < lastNode {
+			nodeHash = m.hasher.HashChildren(nodeHash, proof[proofIndex])
+			proofIndex++
+		} else {
+			// the sibling does not exist and the parent is a dummy copy; do nothing.
+		}
+		nodeIndex = parent(nodeIndex)
+		lastNode = parent(lastNode)
+	}
+	if proofIndex != len(proof) {
+		return nil, fmt.Errorf("invalid proof, expected %d components, but have %d", proofIndex, len(proof))
+	}
+	return nodeHash, nil
+}
+
+// VerifyConsistencyProof checks that the passed in consistency proof is valid between the passed in tree snapshots.
+func (m LogVerifier) VerifyConsistencyProof(snapshot1, snapshot2 int64, root1, root2 []byte, proof [][]byte) error {
+	if snapshot1 > snapshot2 {
+		return fmt.Errorf("snapshot1 (%d) > snapshot2 (%d)", snapshot1, snapshot2)
+	}
+	if snapshot1 == snapshot2 {
```

Comment:
Could pull out the snapshot1 = snapshot2 case as a private helper method. Use your judgment whether it adds clarity or not.